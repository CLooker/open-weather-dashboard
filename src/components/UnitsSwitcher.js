import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default class UnitsSwitcher extends Component {
  static propTypes = {
    name: PropTypes.string,
    setUnits: PropTypes.func.isRequired,
    units: PropTypes.string.isRequired
  };

  handleClick = e => this.props.setUnits(e.target.value.split('')[1]);

  render() {
    const { units } = this.props;
    return (
      <div className="unit-switcher">
        <div>
          <input
            type="button"
            value="&deg;C"
            style={{
              opacity: units === 'F' ? '.5' : '1'
            }}
            onClick={this.handleClick}
          />
          <input
            type="button"
            value="&deg;F"
            style={{
              opacity: units === 'C' ? '.5' : '1'
            }}
            onClick={this.handleClick}
          />
        </div>
      </div>
    );
  }
}
