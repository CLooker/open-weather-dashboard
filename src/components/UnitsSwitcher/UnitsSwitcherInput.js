import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { connect } from 'react-redux';
import { getUnits, setUnits } from '../../redux/modules/units';

console.log(getUnits);

const UnitsSwitcherInputWrapper = styled.input`
  width: 50px;
  opacity: ${props => (props.units === props.value ? '1' : '0.5')};
`;

class UnitsSwitcherInput extends PureComponent {
  static propTypes = {
    units: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
    setUnits: PropTypes.func.isRequired
  };

  render() {
    const { setUnits, units, value } = this.props;
    console.log(this.props);

    return (
      <UnitsSwitcherInputWrapper
        type='button'
        units={units}
        value={value}
        onClick={setUnits}
      />
    );
  }
}

export default connect(
  getUnits,
  { setUnits }
)(UnitsSwitcherInput);
