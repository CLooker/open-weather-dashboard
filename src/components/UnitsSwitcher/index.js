import UnitsSwitcher from './UnitsSwitcher';

export default UnitsSwitcher;
