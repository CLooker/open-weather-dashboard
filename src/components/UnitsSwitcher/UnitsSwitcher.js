import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { farenheit } from '../../utils';
import UnitsSwitcherInput from './UnitsSwitcherInput';

export default class UnitsSwitcher extends Component {
  static propTypes = {
    setUnits: PropTypes.func.isRequired,
    units: PropTypes.string.isRequired
  };

  handleClick = e =>
    this.props.setUnits(e.target.value === farenheit ? 'F' : 'C');

  render() {
    const { units } = this.props;
    return (
      <div className='unit-switcher'>
        <div>
          <UnitsSwitcherInput
            dashboardUnits={units}
            inputUnits={'F'}
            handleClick={this.handleClick}
          />
          <UnitsSwitcherInput
            dashboardUnits={units}
            inputUnits={'C'}
            handleClick={this.handleClick}
          />
        </div>
      </div>
    );
  }
}
