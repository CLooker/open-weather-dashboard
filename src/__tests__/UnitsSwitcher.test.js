import React from 'react';
import { mount, shallow } from 'enzyme';
import UnitsSwitcher from '../components/UnitsSwitcher';

describe('UnitsSwitcher', () => {
  const props = {
    name: 'Chicago',
    units: 'F',
    setUnits: jest.fn()
  };

  let component, mountedComponent;

  beforeEach(() => {
    component = shallow(<UnitsSwitcher {...props} />);
    mountedComponent = mount(<UnitsSwitcher {...props} />);
  });

  afterEach(() => jest.restoreAllMocks());

  describe('initialization', () => {
    it('should render correctly', () => {
      expect(component).toMatchSnapshot();
    });

    it('should have certain props', () => {
      const actual = component.instance().props;
      const expected = props;
      expect(actual).toEqual(props);
    });

    describe('input style based on `units`', () => {
      describe('input style when `units` is "F"', () => {
        test('that the second input has the correct style', () => {
          const actual = mountedComponent
            .find('div input')
            .last()
            .props().style;
          const expected = {
            opacity: '.5'
          };
          expect(actual).toEqual(expect.objectContaining(expected));
        });

        test('that the first input has the correct style', () => {
          const actual = mountedComponent
            .find('div input')
            .first()
            .props().style;
          const expected = {
            opacity: '1'
          };
          expect(actual).toEqual(expect.objectContaining(expected));
        });
      });

      describe('input style when `units` is "C"', () => {
        const localProps = {
          ...props,
          units: 'C'
        };
        const localMountedComponent = mount(<UnitsSwitcher {...localProps} />);

        test('that the first input has the correct style', () => {
          const actual = localMountedComponent
            .find('input')
            .first()
            .props().style;
          const expected = {
            opacity: '.5'
          };
          expect(actual).toEqual(expect.objectContaining(expected));
        });

        test('that the second input has the correct style', () => {
          const actual = localMountedComponent
            .find('input')
            .last()
            .props().style;
          const expected = {
            opacity: '1'
          };
          expect(actual).toEqual(expect.objectContaining(expected));
        });
      });
    });
  });

  describe('method', () => {
    describe('`handleClick`', () => {
      let mockSetUnits, mockValue, localProps, localMountedComponent;

      beforeEach(() => {
        mockSetUnits = jest.fn();
        mockValue = 'ABC';
        localProps = {
          ...props,
          setUnits: mockSetUnits
        };
        localMountedComponent = mount(<UnitsSwitcher {...localProps} />);
        localMountedComponent
          .find('input')
          .first()
          .simulate('click', {
            target: {
              value: mockValue
            }
          });
      });

      it('should call the correct function when called', () => {
        expect(mockSetUnits).toHaveBeenCalled();
      });

      it('should pass the correct argument', () => {
        const actual = mockSetUnits.mock.calls[0][0];
        const expected = mockValue.split('')[1];
        expect(actual).toBe(expected);
      });
    });
  });
});
