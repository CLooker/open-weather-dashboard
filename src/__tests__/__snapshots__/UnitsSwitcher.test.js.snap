// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnitsSwitcher initialization should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UnitsSwitcher
    name="Chicago"
    setUnits={[MockFunction]}
    units="F"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div>
        <input
          onClick={[Function]}
          style={
            Object {
              "opacity": ".5",
            }
          }
          type="button"
          value="°C"
        />
        <input
          onClick={[Function]}
          style={
            Object {
              "opacity": "1",
            }
          }
          type="button"
          value="°F"
        />
      </div>,
      "className": "unit-switcher",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            onClick={[Function]}
            style={
              Object {
                "opacity": ".5",
              }
            }
            type="button"
            value="°C"
          />,
          <input
            onClick={[Function]}
            style={
              Object {
                "opacity": "1",
              }
            }
            type="button"
            value="°F"
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onClick": [Function],
            "style": Object {
              "opacity": ".5",
            },
            "type": "button",
            "value": "°C",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onClick": [Function],
            "style": Object {
              "opacity": "1",
            },
            "type": "button",
            "value": "°F",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div>
          <input
            onClick={[Function]}
            style={
              Object {
                "opacity": ".5",
              }
            }
            type="button"
            value="°C"
          />
          <input
            onClick={[Function]}
            style={
              Object {
                "opacity": "1",
              }
            }
            type="button"
            value="°F"
          />
        </div>,
        "className": "unit-switcher",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              onClick={[Function]}
              style={
                Object {
                  "opacity": ".5",
                }
              }
              type="button"
              value="°C"
            />,
            <input
              onClick={[Function]}
              style={
                Object {
                  "opacity": "1",
                }
              }
              type="button"
              value="°F"
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "onClick": [Function],
              "style": Object {
                "opacity": ".5",
              },
              "type": "button",
              "value": "°C",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "onClick": [Function],
              "style": Object {
                "opacity": "1",
              },
              "type": "button",
              "value": "°F",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
